addons/purchase_stock/models/account_move_line.py

```diff
@@ -124,3 +130,251 @@ def _prepare_in_invoice_svl_vals(self, layers_correction):
+                if initial_layer:
+                    # `layer` is a return. We will cancel the qty to invoice of the returned layer
+                    # /!\ we will cancel the qty not yet invoiced only
+                    initial_layer_remaining_qty = qty_to_invoice_per_layer[initial_layer][1]
+                    common_qty = min(initial_layer_remaining_qty, total_layer_qty_to_invoice)
+                    qty_to_invoice_per_layer[initial_layer][0] -= common_qty
+                    qty_to_invoice_per_layer[initial_layer][1] -= common_qty
+                    total_layer_qty_to_invoice = max(0, total_layer_qty_to_invoice - common_qty)
+                if float_compare(total_layer_qty_to_invoice, 0, precision_rounding=product_uom.rounding) > 0:
+                    qty_to_invoice_per_layer[layer] = [total_layer_qty_to_invoice, total_layer_qty_to_invoice]
+            else:
+                invoice = aml.move_id
+                aml_qty = aml.product_uom_id._compute_quantity(aml.quantity, product_uom)
+                if aml.is_refund:
+                    reversed_invoice = aml.move_id.reversed_entry_id
+                    if reversed_invoice:
+                        sign = -1
+                        impacted_invoice = reversed_invoice
+                        # it's a refund, therefore we can only consume the quantities invoiced by
+                        # the initial invoice (`reversed_invoice`)
+                        layers_to_consume = []
+                        for layer in layers:
+                            remaining_invoiced_qty = layers_and_invoices_qties[(layer, reversed_invoice)][1]
+                            layers_to_consume.append((layer, remaining_invoiced_qty))
+                    else:
+                        # the refund has been generated because of a stock return, let's find and use it
+                        sign = 1
+                        layers_to_consume = []
+                        for layer in qty_to_invoice_per_layer:
+                            if layer.stock_move_id._is_out():
+                                layers_to_consume.append((layer, qty_to_invoice_per_layer[layer][1]))
+                else:
+                    # classic case, we are billing a received quantity so let's use the incoming SVLs
+                    sign = 1
+                    layers_to_consume = []
+                    for layer in qty_to_invoice_per_layer:
+                        if layer.stock_move_id._is_in():
+                            layers_to_consume.append((layer, qty_to_invoice_per_layer[layer][1]))
+                    impacted_invoice = False
+                while float_compare(aml_qty, 0, precision_rounding=product_uom.rounding) > 0 and layers_to_consume:
+                    layer, total_layer_qty_to_invoice = layers_to_consume[0]
+                    layers_to_consume = layers_to_consume[1:]
+                    if float_is_zero(total_layer_qty_to_invoice, precision_rounding=product_uom.rounding):
```

Comment:
But it could become 0
For instance with test `test_pdiff_and_credit_notes`, 
https://github.com/odoo/odoo/blob/56876bb72d68d46ef25efbad44f83daddba233cb/addons/purchase_stock/tests/test_stockvaluation.py#L1938-L1943
When posting BILL03, `layers_to_consume` still contain the first layer, which is fully invoiced, so the quantity will be zero
We still add that layer in `layers_to_consume` because we just check that this is an incoming layer, we don't check the quantity:
https://github.com/odoo/odoo/blob/56876bb72d68d46ef25efbad44f83daddba233cb/addons/purchase_stock/models/account_move_line.py#L230-L235

As there could be some other cases (for instance with refunds), I added the check on the quantity at the end (i.e., where it is now, so we are sure we cover all possible cases)